### PR TITLE
Allow plugin rendering logic to exist outside Graphiql

### PR DESCRIPTION
**Pull request by ismail-syed**
This PR allows for the rendering logic of the plugin view to be done from outside the GraphiQL core codebase. 

screen shot 2018-02-11 at 4 55 00 pm

### CODE REVIEW

File: test/index.html [hunk 1]
```diff
@@ -126,6 +128,17 @@
         });
       }
 
+      const viewer = function(results) {
```
Comment (ismail-syed) [inline test/index.html:131]: This is a function that takes in a `results` param. This params is the results strings in the GraphiQL results page. 
The plugin function can process the results string and return whatever UI it want's rendered.

File: test/index.html [hunk 2]
```diff
@@ -26,6 +26,8 @@
     <script src="./vendor/react-15.4.2.js"></script>
     <script src="./vendor/react-dom-15.4.2.js"></script>
 
+    <script src="https://cdnjs.cloudflare.com/ajax/libs/babel-core/5.8.23/browser.js"></script>
```
Comment (ismail-syed) [inline test/index.html:29]: This is just for the initial proof of concept PR. It's importing `babel-core` so the scripts below can handle jsx syntax

File: src/components/PluginsPane.js [hunk 3]
```diff
@@ -49,20 +52,22 @@ export class PluginsPane extends React.Component {
...
-      <TabList>{tabs}</TabList>
-    )
+    return <TabList>{tabs}</TabList>;
   }
 
   tabPanels() {
-    const tabs = this.props.plugins.map((plugin, id) => {
+    const tabs = this.props.plugins.map((aPlugin, id) => {
+      const { plugin } = aPlugin;
+      const viewer = plugin && plugin(this.props.value);
```
Comment (ismail-syed) [inline src/components/PluginsPane.js:63]: This is where we call the plugin function and pass it the results string. 
```
plugin(this.props.value)
```

File: src/components/GraphiQL.js [hunk 4]
```diff
@@ -382,13 +382,14 @@ export class GraphiQL extends React.Component {
                 editorTheme={this.props.editorTheme}
                 ResultsTooltip={this.props.ResultsTooltip}
               />
-              <GraphiQL.Footer>
-                {this.props.plugins
-                  ? <PluginsPane
+              {this.props.plugins
+                ? <GraphiQL.Footer>
```
Comment (DerekStride) [inline src/components/GraphiQL.js:386]: So if there's no plugins we don't render the plugin pane?

File: src/components/PluginsPane.js [hunk 5]
```diff
@@ -4,14 +4,17 @@ import PropTypes from 'prop-types';
...
-  }];
+  static propTypes = [
+    {
+      plugins: PropTypes.arrayOf(
+        PropTypes.shape({
+          title: PropTypes.string,
+          plugin: PropTypes.func.isRequired,
+        }),
+      ),
+      value: PropTypes.string,
```
Comment (DerekStride) [inline src/components/PluginsPane.js:15]: What is this?

File: src/components/GraphiQL.js (hunk 4 above)
Comment (ismail-syed) [reply src/components/GraphiQL.js:386]: Yup, we just render the original footer like they do here: https://github.com/graphql/graphiql/blob/master/src/components/GraphiQL.js#L377

File: src/components/PluginsPane.js (hunk 5 above)
Comment (ismail-syed) [reply src/components/PluginsPane.js:15]: This is a prop that accepts the results value as a string. It's was sent to the current `ResultsViewer` component that renders the results json obj. 
We're passing it into the PluginsPane to share it with all the plugins. It's used here: 
https://github.com/GraphQL-Query-Planner/graphiql/blob/extract-plugin/src/components/GraphiQL.js#L389
A few lines above you can see the same pattern is used to share the results with the `ResultViewer`

**Review by DerekStride — APPROVED** [review]

**Pull request merged**